### PR TITLE
Fix tests to not hang on Windows with bad .py assoc

### LINKED ISSUE #4053: testsuite won't run on Windows with VS Code installed
Mostly filed for recording purposes - if others have more defails please add.

On systems with Visual Studio Code installed, with the Python extension and pylance (which is now the favored engine), Code "steals" the .py file association. This has the effect that tests which try to execute a script directly (usually via hand-crafted shebang, which typically isn't very useful on Windows anyway) end up not running, but opening an instance of Code with the my_python script ready for editing.  Because of the strange way scons-time tests work, this means the scons-time tests aren't usable.  

Further, in at least three instances, there's a fatal javascript exception on the screen, but at the moment I don't have any way to associate that with the test(s) that trigger it. **Update**: this comes from `test/sconsign/script/SConsignFile.py`, apparently. The error has something to wo with `%USERPROFILE%` underfined, and JavaScript is involved because it's VS Code firing this up again...  this may be related to the recent claim that we're executing Windows processes with too limited a set of environment vars.

In the end, probably due to the javascript-triggered error, the thread lock is never released, so the testsuite never "finishes".  

After manually killing things, we get an exception on unraisablehook: https://docs.python.org/3/library/sys.html#sys.unraisablehook

The runtest Tee class apparently needs to implement a flush method, as I've seen that complaint as well.

This is all fairly recent, observed now on two Windows systems, at least one of them which has successfully run the SCons testsuite without odd problems in the not too distant past.  We are not seeing these problems on the CI runs using AppVeyor.

 Not yet sure if this is an indication we're doing dodgy things and only now getting punished for it, or if VS Code is overstepping its bounds.

### PR DESCRIPTION
**Pull request by mwichmann**
For systems where the association for .py files is not to an actual Python interpreter, those few cases where we need to run a Python script directly as a program don't work.  This could be because the association was never set up, or because some other program (e.g. Visual Studio Code) has taken it over. In some cases may appear to "hang" because the alternate program is waiting for user interaction

`runtest.py` now has a mechanism to check (thanks to Brett Cannon for providing this incantation). It isn't super precise (looks for the substring "py" in the queried association), but should work out.  It sets an environment variable which the test framework can read and as a result set a flag which individual tests can read.

Two tests in scons-time which had previously been set to skip-if-win32 now look at this flag instead.  Three tests in sconsign now also look at this flag. This allows a clean run on my dev box with VS Code having taken over the .py association.

Various things can break if the environment used to fire off Windows processes doesn't contain `%UserProfile%`. Added this to the short list of passthrough env vars.  Apparently an environment without this value is now considered invalid (it blew up the erroneously launched VS Code, but we've apparently been lucky it hasn't blown up more things - believe there was also a report of a problem with the Visual Studio setup scripts).

A little extra cleanup:
- a couple of Py2-isms were cleaned out (`Script/Main.py` and in the test framework)
- The paths to look for site-scons were rewritten (part of this was another Py2-ism), and the system path changed a bit - the old path is still checked, and the manpage updated to reflect this.
- `runtest.py` dropped the unused whereis functions.
- the three sconsign tests now use f-string formatting, mostly as an experiment to see how easy it is to convert.

Fixes #4053

Signed-off-by: Mats Wichmann <mats@linux.com>

## Contributor Checklist:

* [X] I have created a new test or updated the unit tests to cover the new/changed functionality.
* [X] I have updated `CHANGES.txt` (and read the `README.rst`)
* [X] I have updated the appropriate documentation

### CODE REVIEW

File: CHANGES.txt [hunk 1]
```diff
@@ -13,6 +13,14 @@ RELEASE  VERSION/DATE TO BE FILLED IN LATER
 
         - Whatever John Doe did.
 
+  From Mats Wichmann:
+    - Tweak definition of default site_scons paths on Windows to conform
```
Comment (bdbaddog) [inline CHANGES.txt]: Please list out the changes explicitly here and in RELEASE.txt.

File: CHANGES.txt [hunk 2]
```diff
@@ -13,6 +13,14 @@ RELEASE  VERSION/DATE TO BE FILLED IN LATER
...
+      to conventions, drop a Py2 workaround, and pick as better "system" path
+      (old one remains supported).
+    - Fix testsuite to work on Windows systems where there is no usable
```
Comment (bdbaddog) [inline CHANGES.txt:22]: Add a note that Visual Studio Code is the usual suspect when this association get's broken?

File: CHANGES.txt (hunk 1 above)
Comment (mwichmann) [reply CHANGES.txt]: Added the new path to RELEASE.txt, and clarified the description in CHANGES.txt.

File: CHANGES.txt (hunk 2 above)
Comment (mwichmann) [reply CHANGES.txt:22]: Added.

File: doc/man/scons.xml [hunk 3]
```diff
@@ -1729,27 +1729,28 @@ Also suppresses SCons status messages.</para>
   </varlistentry>
 
   <varlistentry id="opt-site-dir">
-  <term><option>--site-dir=<replaceable>dir</replaceable></option></term>
+  <term><option>--site-dir=<replaceable>path</replaceable></option></term>
```
Comment (bdbaddog) [inline doc/man/scons.xml:1732]: It should be a dir right? it's not a path list, and folks may confuse it as such?

File: runtest.py [hunk 4]
```diff
@@ -258,31 +258,42 @@ def flush(self, data):
...
+    ASSOCF_INIT_IGNOREUNKNOWN = 0x00000400
+    ASSOCSTR_COMMAND = 1
+    ASSOCSTR_EXECUTABLE = 2
+    E_POINTER = ctypes.c_long(0x80004003).value
+
+    def get_template_command(filetype, verb=None):
+        flags = ASSOCF_INIT_IGNOREUNKNOWN | ASSOCF_NOTRUNCATE
+        assoc_str = ASSOCSTR_COMMAND
+        cch = ctypes.c_ulong(260)
+        while True:
```
Comment (bdbaddog) [inline runtest.py:285]: Why `while True` ?

File: runtest.py [hunk 5]
```diff
@@ -528,6 +539,19 @@ def footer(self, f):
...
+    # Windows doesn't support "shebang" lines directly (the Python launcher
+    # and Windows Store version do, but you have to get them launched first)
+    # so to directly launch a script we depend on an assoc for .py to work.
+    # Some systems may have none, and in some cases IDE programs take over 
+    # the assoc.  Detect this so the small number of tests affected can skip.
+    try:
+        python_assoc = get_template_command('.py')
+    except OSError:
+        python_assoc = None
+    if not python_assoc or "py" not in python_assoc:
```
Comment (bdbaddog) [inline runtest.py:552]: What's python_assoc?

Comment (mwichmann) [reply runtest.py:552]: made-up local variable for the looked-up association to the Python filetype.

File: runtest.py (hunk 4 above)
Comment (mwichmann) [reply runtest.py:285]: Beats me. I got an updated snip just now I'll check if it still does that.

File: runtest.py (hunk 5 above)
Comment (bdbaddog) [reply runtest.py:552]: ahh.. always have coffee before reviewing.. ;)

**Pull request merged**